tests/azure_iot_hub_client_ut.c

```diff
@@ -381,6 +475,690 @@ static void testAzureIoTHubClient_SendTelemetry_Success( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTSuccess );
+    assert_int_equal( AzureIoTHubClient_ProcessLoop( &xTestIoTHubClient,
+                                                     1234 ),
+                      eAzureIoTHubClientSuccess );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_CloudMessageSubscribe_InvalidArgFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    /* Fail SubscribeCloudToDevice when client is NULL */
+    assert_int_equal( AzureIoTHubClient_SubscribeCloudToDeviceMessage( NULL,
+                                                                       prvTestCloudMessage,
+                                                                       NULL, ( uint32_t ) -1 ),
+                      eAzureIoTHubClientInvalidArgument );
+
+    /* Fail SubscribeCloudToDevice when function callback is NULL  */
+    assert_int_equal( AzureIoTHubClient_SubscribeCloudToDeviceMessage( &xTestIoTHubClient,
+                                                                       NULL, NULL, ( uint32_t ) -1 ),
+                      eAzureIoTHubClientInvalidArgument );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_CloudMessageSubscribe_SendFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Subscribe, eAzureIoTMQTTSendFailed );
+    assert_int_equal( AzureIoTHubClient_SubscribeCloudToDeviceMessage( &xTestIoTHubClient,
+                                                                       prvTestCloudMessage,
+                                                                       NULL, ( uint32_t ) -1 ),
+                      eAzureIoTHubClientSubscribeFailed );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_CloudMessageSubscribe_ReceiveFailure( void ** state )
+{
+    AzureIoTHubClient_t xTestIoTHubClient;
+
+    ( void ) state;
+
+    prvSetupTestIoTHubClient( &xTestIoTHubClient );
+
+    will_return( AzureIoTMQTT_Subscribe, eAzureIoTMQTTSuccess );
+    will_return( AzureIoTMQTT_ProcessLoop, eAzureIoTMQTTRecvFailed );
+    assert_int_equal( AzureIoTHubClient_SubscribeCloudToDeviceMessage( &xTestIoTHubClient,
+                                                                       prvTestCloudMessage,
+                                                                       NULL, ( uint32_t ) -1 ),
+                      eAzureIoTHubClientFailed );
+}
+/*-----------------------------------------------------------*/
+
+static void testAzureIoTHubClient_CloudMessageSubscribe_Success( void ** state )
```

Comment:
Could we add a new test where AzureIoTHubClient_SubscribeCloudToDeviceMessage (and analogous other subscribes) are called multiple times, successfully.  It's possible an app could do this - like getting confused on state - and since MQTT protocol and our SDK allow this, we should make it works.